host/xtest/asym_perf.c

```diff
@@ -0,0 +1,962 @@
+
+		xtest_add_attr(&param_count, params, TEE_ATTR_DH_PRIME,
+			       keygen_dh_p, BITS_TO_BYTES(width_bits));
+		xtest_add_attr(&param_count, params, TEE_ATTR_DH_BASE,
+			       keygen_dh_g, DH_G_SIZE);
+		break;
+	case ALGO_ECDSA:
+	case ALGO_ECDH:
+		curve_id = get_curve_id(width_bits);
+		if (curve_id == TEE_CRYPTO_ELEMENT_NONE)
+			return -1;
+		xtest_add_attr_value(&param_count, params, TEE_ATTR_ECC_CURVE,
+				     curve_id, 0);
+		break;
+	case ALGO_X25519:
+		width_bits = WIDTH_BITS_25519;
+		break;
+	default:
+		break;
+	}
+
+	prepare_obj(width_bits, main_algo);
+
+	res = pack_attrs(params, param_count, &buf, &blen);
+	CHECK(res, "pack_attrs", goto out;);
+
+	if (mode == MODE_GENKEYPAIR) {
+		if (blen)
+			prepare_attrs(buf, blen);
+
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_PROCESS_GEN_KEYPAIR;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+						 TEEC_NONE, TEEC_NONE,
+						 TEEC_NONE);
+		op.params[0].value.a = width_bits;
+		op.params[0].value.b = l;
+	} else {
+		prepare_keypair(width_bits, buf, blen, mode, algo, main_algo);
+
+		if (mode == MODE_SIGN || mode == MODE_VERIFY)
+			prepare_hash(size, main_algo);
+
+		if (main_algo == ALGO_RSA && algo >= RSASSA_PKCS1_PSS_MGF1_SHA1) {
+			params[0].attributeID = TEE_ATTR_RSA_PSS_SALT_LENGTH;
+			params[0].content.value.a = salt_len;
+			params[0].content.value.b = 0;
+			param_count = 1;
+			res = pack_attrs(params, param_count, &buf, &blen);
+			CHECK(res, "pack_attrs", goto out;);
+		}
+
+		if (blen)
+			prepare_attrs(buf, blen);
+
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_PROCESS;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+						 TEEC_MEMREF_PARTIAL_INPUT,
+						 TEEC_MEMREF_PARTIAL_OUTPUT,
+						 TEEC_NONE);
+
+		op.params[0].value.a = l;
+		op.params[0].value.b = mode;
+
+		switch (mode) {
+		case MODE_ENCRYPT:
+			op.params[1].memref.parent = &in_shm;
+			op.params[1].memref.size = size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		case MODE_SIGN:
+			op.params[1].memref.parent = &hash_shm;
+			op.params[1].memref.size = hash_shm.size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		case MODE_DECRYPT:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[1].memref.parent = &out_shm;
+			op.params[1].memref.size = out_shm.size;
+			op.params[2].memref.parent = &in_shm;
+			op.params[2].memref.size = size;
+			break;
+		case MODE_VERIFY:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[1].memref.parent = &hash_shm;
+			op.params[1].memref.size = hash_shm.size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		default:
+			fprintf(stderr, "Unexpected mode value\n");
+			exit(1);
+		}
+	}
+
+	if (warmup)
+		do_warmup(warmup);
+
+	while (n-- > 0) {
+		if (mode == MODE_ENCRYPT && main_algo == ALGO_RSA) {
+			/* Make sure the N > M */
```

Comment:
Suggestion: s/`Make sure the N > M`/`Ensure N > M`/
Same suggestion for line 439.